TRIA-117 Allow superuser to export mobile app setup for a user

bramj opened the pull request:
This PR adds a feature that allows a superuser to download a .zip file for a given user and project.
This zip file contains all the files needed to do the complete setup of the mobile app. It contains the user's profile, an access token and .json files for all resources such as the org unit pyramid and the entities, based on the enabled feature flags.

## Self proofreading checklist

- [x] Did I use eslint and black formatters
- [x] Is my code clear enough and well documented
- [x] Are my typescript files well typed
- [x] New translations have been added or updated if new strings have been introduced in the frontend
- ~My migrations file are included~
- [x] Are there enough tests
- [ ] Documentation has been included (for new feature)

## Doc
- Tell us where the doc can be found (docs folder, wiki, in the code...)

## Changes

- Added an action (icon button) on the users list if the authorized user is a super user
- Generating the export will:
  1. Launch a Task
  2. Poll the task for it's status and progress
  3. When the task has completed successfully, request a signed url (S3) for the generated file
  4. Display the download button
- The task will call all API endpoints and write the responses to files in a temporary directory. When done it adds all files to an archive and uploads to S3.

## How to test

1. Set `DNS_DOMAIN` Env variable to your local machine. Note: I had to use `ngrok` to get it to work because it didn't find my localhost from inside the docker container. Using `ngrok` was the faster solution.
2. Launch the tasks_worker:
  ```
  docker-compose run --rm iaso bash
  python manage.py tasks_worker
  ```
3. Make sure you're a super user and use the dashboard to generate an export

## Print screen / video



https://github.com/BLSQ/iaso/assets/2719739/4774f879-a986-4714-9eaf-584e173e3ec3

bramj committed:
TRIA-117 Move IasoClient to /iaso for wider use

bramj committed:
TRIA-117 Iaso API client wrapper: spit auth from init

This allows us to have a client without authentication, or to
authenticate using user/pass or directly with a token.

bramj committed:
TRIA-117 Task to generate mobile app setup files first version

bramj committed:
TRIA-117 Add Django test for mobile app export

bramj committed:
TRIA-117 Refactor export task for better readability

bramj committed:
TRIA-117 Mobile app export working version

bramj committed:
TRIA-117 Add reports and small fixes on mobile app setup export

bramj committed:
TRIA-117 Add zip export to frontend first version

bramj committed:
TRIA-117 Improve export modal

bramj committed:
TRIA-117 Improve the dialog, first version of task creation working

bramj committed:
TRIA-117 Bring back Son's TaskMonitor

Very useful here to monitor the running tasks to retrieve it's
progress.

bramj committed:
TRIA-117 Finish progress bar up to download button...

with S3 presigned URL.

bramj committed:
TRIA-117 Improve export file modal copy + error handling

bramj committed:
TRIA-117 Fix: Presigned S3 URL was not working

bramj committed:
TRIA-117 Add submodule to show Task details

bramj committed:
TRIA-117 Fix SERVER_URL and double call to presigned url

bramj commented:
General thoughts:
- Personally I find the default MUI button less than great, maybe we can use the "contained" variant everywhere? https://mui.com/material-ui/react-button/#basic-button
- I think the title of our dialog is missing some style (e.g. the bold part): https://mui.com/material-ui/react-dialog/#alerts It's a detail, but it does look better on the MUI docs. But I'm not sure how to fix it

bramj committed:
TRIA-117 Fix CI

beygorghor commented on hat/assets/js/apps/Iaso/domains/tasks/components/TaskDetails.tsx:
```diff
@@ -0,0 +1,62 @@
+    useEffect(() => {
+        if (taskHasDownloadableFile) {
+            getRequest(`/api/tasks/${task.id}/presigned-url/`).then(resp =>
+                setPresignedUrl(resp.presigned_url),
```
Why not react-query hook enabled if task.id and watching it ?

beygorghor commented on hat/assets/js/apps/Iaso/domains/tasks/index.js:
```diff
@@ -108,6 +109,10 @@ const Tasks = ({ params }) => {
                     extraProps={{
                         loading: isLoading,
+                        SubComponent: task => {
+                            console.log('task', task);
```
🗑️

beygorghor commented on hat/assets/js/apps/Iaso/domains/users/components/ExportMobileAppSetupDialog.tsx:
```diff
@@ -0,0 +1,265 @@
+    props: LinearProgressProps & { value: number },
+) {
+    return (
+        <Box sx={{ display: 'flex', alignItems: 'center' }}>
```
this is a FunctionComponent

beygorghor replied:
also you can use proper sxStyles for it

bramj replied:
> this is a FunctionComponent

Sorry, but what do you mean by this comment? Not sure what I should change

beygorghor commented on hat/assets/js/apps/Iaso/domains/users/components/ExportMobileAppSetupDialog.tsx:
```diff
@@ -0,0 +1,265 @@
+                            multi={false}
+                            label={MESSAGES.project}
+                            options={selectedUser.projects.map(project => {
+                                return {
```
maybe a useMemo for this

beygorghor reviewed (changes requested):
Some small changes

bramj committed:
TRIA-117 Remove console.log and use React Query for task details

bramj committed:
TRIA-117 Don't export access_token and user profile

bramj committed:
TRIA-117 Download forms versions and attachements

And add it all to the correct folders and create the zipfile
accordingly.

bramj committed:
TRIA-117 Only save paginated calls as *-1.json etc.

bramj committed:
TRIA-117 Small frontend code improvements based on PR feedback

bramj committed:
TRIA-117 Rewrite file urls for links to XML files

bramj committed:
TRIA-117 Fix export mobile app test

bramj committed:
TRIA-117 Encrypt zip before uploading to S3

bramj committed:
TRIA-117 Make export fail if one of the API calls fails

For example the call to entities fails with a 403 due to insufficient
permissions.

bramj committed:
TRIA-117 Allow user to choose a password to encrypt mobile setup

bramj committed:
TRIA-117 Add some documentation for the encryption functions

bramj committed:
TRIA-117 Fix: Improve parsing of URLs

AWS signed URLs gave issues because of their query params.

bramj committed:
TRIA-117 Fix: Reports is a full URL, not a relative path

bramj committed:
TRIA-117 Fix: Form attachment manifest URL

bramj committed:
TRIA-117 Fix: Don't send auth header when S3 url

mathvdh commented on iaso/tasks/export_mobile_app_setup_for_user.py:
```diff
@@ -0,0 +1,282 @@
+        the_task.report_progress_and_stop_if_killed(progress_value=the_task.progress_value + 1)
+
+    s3_object_name = _compress_and_upload_to_s3(tmp_dir, export_name, password)
+    print("s3_object_name ")
```
I know we are in a task but do we want to keep "print" ? I mean don't we have to use the logging system instead ? 🤔

bramj replied:
You're right, this is actually a debug leftover, good catch

mathvdh commented on iaso/tasks/export_mobile_app_setup_for_user.py:
```diff
@@ -0,0 +1,282 @@
+
+        with open(os.path.join(tmp_dir, "forms", filename), mode="wb") as f:
+            f.write(response.content)
+
```
This could be eventually put in a function and reused on lines 246 -> 255

bramj replied:
👍 done, I think I was a bit lazy because of the small difference with the url/path in case of a non S3 url

mathvdh reviewed (changes requested):
I think this is really great ! 👍🏼 
I have small comments and suggestions below, but otherwise it's really good.
Some questions I have : 
I like the API_CALLS file ... it's really SMART.
I'm not sure how it works on import in mobile app (like I can' t really test it).
BUT would it be possible in the future to import such a zip file locally ? Like from server A of iaso to server B (if we would add like a screen where you can select with check boxes ;; export this and that ;; then import them on another server it would be a killer feature

bramj committed:
TRIA-117 Refac: PR feedback, extract function (DRY up)

bramj commented:
> I think this is really great ! 👍🏼 I have small comments and suggestions below, but otherwise it's really good. Some questions I have : I like the API_CALLS file ... it's really SMART. I'm not sure how it works on import in mobile app (like I can' t really test it). BUT would it be possible in the future to import such a zip file locally ? Like from server A of iaso to server B (if we would add like a screen where you can select with check boxes ;; export this and that ;; then import them on another server it would be a killer feature

Thanks for the kind words and the review :) You mean if we could import such a zip on the backend of another server? I suppose there's not really a reason why not 🤔 Though if we would add such a feature maybe it would make more sense to enter an API endpoint (pointing to server A) on your server B and have it pull in the data with a sort of setuper.

But with this zip file the data is specific to the mobile app, so I suppose you'd also miss a lot of data like fields on the OrgUnits that are simply not in the mobile payload.

Do you think there's a demand for a server A to server B import?

bramj commented:
@mathvdh Okay for merge
